aiohttp/web_request.py

```diff
@@ -148,7 +148,7 @@ def __init__(
         self,
         message: RawRequestMessage,
         payload: StreamReader,
-        protocol: "RequestHandler",
+        protocol: "RequestHandler[BaseRequest]",
```

Comment:
Without looking at the rest of the PR yet, this stands out to me as something where the complexity magnifies.

If we subclass BaseRequest, then we'll surely call super().__init__(), which is then requiring a protocol that can handle any BaseRequest. So, in what situation could you have a protocol that only handles a subclass, given that you can't pass it to a subclass?

Reply:
This type constraint is "really" `RequestHandler[type(self)]`, i.e. the protocol is able to handle the current type. As `type(self)` is logically constrained to be a subclass or `BaseRequest`, all protocols passed into this `__init__` will have to handle the properties of `BaseRequest`.

A subclass may change the annotation for clarity
```python
from __future__ import annotations

class Request(aiohttp.web.BaseRequest):
    def __init__(
        self,
        message: aiohttp.http.RawRequestMessage,
        payload: aiohttp.streams.StreamReader,
        protocol: aiohttp.web.RequestHandler[Request],
        payload_writer: aiohttp.abc.AbstractStreamWriter,
        task: asyncio.Task[None],
    ) -> None:
        super().__init__(
            message, payload, protocol, payload_writer, task, asyncio.get_running_loop()
        )
```

Reply:
> A subclass may change the annotation for clarity

My point is that, that code will produce a type error. If protocol requires `[Request]` then it not compatible with a parameter than requires `[BaseRequest]`. Because if the protocol does something specifically requires attributes that only exist on Request, then it would fail when it received a BaseRequest and therefore can't be passed through the super() call without a type error. In other words, the type annotation would need to be generic as well, rather than using BaseRequest, which is where it gets rather complicated.